// fairmont/lib/string.js
// Generated by CoffeeScript 1.9.1
(function() {
  var assert, describe, ref;

  ref = require("./helpers"), describe = ref.describe, assert = ref.assert;

  describe("String functions", function(context) {
    var blank, camel_case, capitalize, dashed, html_escape, plain_text, title_case, to_string, underscored, w;
    to_string = function(x) {
      return x.toString();
    };
    plain_text = function(string) {
      return string.replace(/^[A-Z]/g, function(c) {
        return c.toLowerCase();
      }).replace(/[A-Z]/g, function(c) {
        return " " + (c.toLowerCase());
      }).replace(/\W+/g, " ");
    };
    context.test("plain_text", function() {
      assert(plain_text("hello-world") === "hello world");
      return assert(plain_text("Hello World") === "hello world");
    });
    capitalize = function(string) {
      return string[0].toUpperCase() + string.slice(1);
    };
    context.test("capitalize", function() {
      return assert(capitalize("hello world") === "Hello world");
    });
    title_case = function(string) {
      return string.toLowerCase().replace(/^(\w)|\W(\w)/g, function(char) {
        return char.toUpperCase();
      });
    };
    context.test("title_case", function() {
      return assert(title_case("hello woRld") === "Hello World");
    });
    camel_case = function(string) {
      return string.toLowerCase().replace(/(\W+\w)/g, function(string) {
        return string.trim().toUpperCase();
      });
    };
    context.test("camel_case", function() {
      return assert(camel_case("Hello World") === "helloWorld");
    });
    underscored = function(string) {
      return plain_text(string).replace(/\W+/g, "_");
    };
    context.test("underscored", function() {
      return assert(underscored("Hello World") === "hello_world");
    });
    dashed = function(string) {
      return plain_text(string).replace(/\W+/g, "-");
    };
    context.test("dashed", function() {
      return assert(dashed("Hello World") === "hello-world");
    });
    html_escape = (function() {
      var entities, map, re;
      map = {
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        '"': '&quot;',
        "'": '&#39;',
        "/": '&#x2F;'
      };
      entities = Object.keys(map);
      re = new RegExp("" + (entities.join('|')), "g");
      return function(string) {
        return string.replace(re, function(s) {
          return map[s];
        });
      };
    })();
    context.test("html_escape", function() {
      return assert.equal(html_escape("<a href='foo'>bar & baz</a>"), "&lt;a href=&#39;foo&#39;&gt;bar &amp; baz&lt;&#x2F;a&gt;");
    });
    w = function(string) {
      return string.trim().split(/\s+/);
    };
    context.test("w", function() {
      return assert((w("one two three")).length === 3);
    });
    blank = function(s) {
      return s.length === 0;
    };
    context.test("blank", function() {
      assert(blank(""));
      return assert(!blank("x"));
    });
    return module.exports = {
      to_string: to_string,
      capitalize: capitalize,
      title_case: title_case,
      camel_case: camel_case,
      underscored: underscored,
      dashed: dashed,
      plain_text: plain_text,
      html_escape: html_escape,
      w: w,
      blank: blank
    };
  });

}).call(this);
