// fairmont/lib/type.js
// Generated by CoffeeScript 1.9.1
(function() {
  var assert, curry, deep_equal, describe, ref, ref1;

  ref = require("./core"), curry = ref.curry, deep_equal = ref.deep_equal;

  ref1 = require("./helpers"), describe = ref1.describe, assert = ref1.assert;

  describe("Type functions", function(context) {
    var instance_of, is_array, is_function, is_object, is_string, is_type, is_value, type;
    type = function(x) {
      return Object.prototype.toString.call(x).slice(8, -1).toLowerCase();
    };
    context.test("type");
    is_type = curry(function(t, x) {
      return type(x) === t;
    });
    context.test("is_type");
    instance_of = curry(function(t, x) {
      return x instanceof t;
    });
    context.test("instance_of");
    is_string = is_type("string");
    is_function = is_type("function");
    is_object = is_type("object");
    is_array = is_type("array");
    is_value = function(x) {
      return x != null;
    };
    return module.exports = {
      deep_equal: deep_equal,
      type: type,
      is_type: is_type,
      instance_of: instance_of,
      is_string: is_string,
      is_function: is_function,
      is_object: is_object,
      is_array: is_array,
      is_value: is_value
    };
  });

}).call(this);
